### ISSUE TITLE
Can I use ObservableColection<T> instead of ChartValues<T> as datasource for series in LiveCharts ?.

### ISSUE THREAD

**Issue by EugeneBuzin** [beto-rodriguez/Live-Charts#284]
I'm beginner in LiveCharts, so I have a question. Is it possible  to use an ObservableCollection<T> instance instead of ChartValues<T> instance as datasource for series in LiveCharts? Thank you very much in advance.

**Comment by beto-rodriguez** [comment]
ChartValues is much more than an array, or an observable collection, it is a specialized collection for the library.

I once heard an opinion that our VMs should not include ChartValues and he got a point, in that case you could always use a converter from any `IList`

    var myArray = new double[] {1,2,3};
    var lineSeries = new LineSeries 
    {
        Values = myArray.AsChartValues()
    }

let me know if I got your point.

**Comment by MeshackMusundi** [comment]
@beto-rodriguez The chart doesn't update smoothly when using a converter. I've attempted to use one in the WPF sample project, in the `ConstantChangesChart` control.
```
public class CollectionChartValuesConverter : IValueConverter
{
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        var data = (value as ObservableCollection<MeasureModel>).AsChartValues<MeasureModel>();
        return data;
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return Binding.DoNothing;
    }
}
```

**Comment by ReedCopsey** [comment]
> ChartValues is much more than an array, or an observable collection, it is a specialized collection for the library.

@beto-rodriguez Reading this while trying to help somebody using your library - looking at the code, your ChartValues are effectively just based on your NoisyCollection collections, which really are just a subset of the change notification functionality provided by INotifyCollectionChanged.  If you were to make the library work against INotifyCollectionChanged instead of NoisyCollection, you could easily make the chart series bindable _and_ keep a similar level of performance.

This would allow people to use your charts in more "pure MVVM" fashion, as the ViewModel layers wouldn't need to know about, or use types, defined within the View.  It'd also open up a lot of other scenarios that only exist when you're using data binding, which currently require horrible hacks to work around (such as using a converter, as mentioned here, which destroys the performance).

(Note that, to keep the perf. the same, you'd probably need to make NoisyCollection use your own implementation of INotifyCollectionChanged which included an AddRange method - but the underlying event handling supports that scenario, even though ObservableCollection doesn't).

**Comment by beto-rodriguez** [comment]
@ReedCopsey this is one of the things I don't like in the current version of the library, the library forces you to use SeriesCollection/ChartValues (that inherits from NoisyCollection), but when you do you can't notify the changes to the UI, this is an old thread but I've had some discussions about it, and I know this is annoying.

Essentially as you stated the solution  is to Implement `INotifyCollectionChanged`.

This will make things cleaner for everyone.

**Comment by ReedCopsey** [comment]
@beto-rodriguez Yes, but unfortunately, that's a violation of MVVM tenants - again, it is forcing you to put "view types" into the VM layer.  I just wanted to mention this to you - hopefully a future version of this library can take that into account and change the requirements to using INCC so things are bindable.

**Comment by beto-rodriguez** [comment]
@ReedCopsey Thank you very much for your feedback, last commit as you suggested implements `INotifyCollectionChanges` and `INotifyPropertychanged` (count and indexer properties), as far as I see everything works correctly now and its a clean solution.

This will be included in the next version of the library.

in case some one is interested:

    <Grid>
        <Grid.RowDefinitions>
            <RowDefinition Height="Auto"></RowDefinition>
            <RowDefinition Height="Auto"></RowDefinition>
            <RowDefinition Height="*"></RowDefinition>
        </Grid.RowDefinitions>
        <Grid.ColumnDefinitions>
            <ColumnDefinition Width="Auto"></ColumnDefinition>
            <ColumnDefinition Width="*"></ColumnDefinition>
        </Grid.ColumnDefinitions>
        <StackPanel Grid.Row="0">
            <Button Click="AddOnClick">Add</Button>
            <Button Click="AddRangeOnClick">Add Range</Button>
            <Button Click="InsertOnClick">Insert</Button>
            <Button Click="InsertRangeOnClick">InsertRange</Button>
            <Button Click="RemoveOnClick">Remove</Button>
            <Button Click="RemoveAtOnClick">RemoveAt</Button>
            <Button Click="ClearOnclick">Clear</Button>
        </StackPanel>
        <TextBlock Grid.Row="1" Grid.Column="0" Text="{Binding Values.Count, StringFormat={}count: {0}}" />
        <ListBox Grid.Row="2" Grid.Column="0" ItemsSource="{Binding Values}">
        </ListBox>
        <lvc:CartesianChart Grid.Row="2" Grid.Column="1">
           <lvc:CartesianChart.Series>
               <lvc:LineSeries Values="{Binding Values}"></lvc:LineSeries>
           </lvc:CartesianChart.Series>
        </lvc:CartesianChart>
    </Grid>

    public partial class MainWindow
    {
        private Vm Vm;

        public MainWindow()
        {
            Vm = new Vm();
            DataContext = Vm;
        }

        private void AddOnClick(object sender, RoutedEventArgs e)
        {
            Vm.Values.Add(new Random().Next(0, 10));
        }

        private void AddRangeOnClick(object sender, RoutedEventArgs e)
        {
            var r = new Random();
            Vm.Values.AddRange(new List<double> {r.Next(0, 10), r.Next(0, 10)});
        }

        private void InsertOnClick(object sender, RoutedEventArgs e)
        {
            if (Vm.Values.Count < 2) return;
            Vm.Values.Insert(1, new Random().Next(0, 10));
        }

        private void InsertRangeOnClick(object sender, RoutedEventArgs e)
        {
            if (Vm.Values.Count < 2) return;
            var r = new Random();
            Vm.Values.InsertRange(1, new List<double> {r.Next(0, 10), r.Next(0, 10)});
        }

        private void RemoveOnClick(object sender, RoutedEventArgs e)
        {
            if (Vm.Values.Count < 2) return;
            Vm.Values.Remove(Vm.Values[Vm.Values.Count - 1]);
        }

        private void RemoveAtOnClick(object sender, RoutedEventArgs e)
        {
            if (Vm.Values.Count < 2) return;
            Vm.Values.RemoveAt(Vm.Values.Count - 1);
        }

        private void ClearOnclick(object sender, RoutedEventArgs e)
        {
            Vm.Values.Clear();
        }
    }

    public class Vm
    {
        public Vm()
        {
            Values = new ChartValues<double>
            {
                1,
                6,
                3,
                8
            };
        }

        public ChartValues<double> Values { get; set; }
    }

**Comment by ReedCopsey** [comment]
@beto-rodriguez This part:

```public class Vm
{
    public Vm()
    {
        Values = new ChartValues<double>
        {
            1,
            6,
            3,
            8
        };
    }

    public ChartValues<double> Values { get; set; }
}```

Ideally, you should be able to use:
```
public class Vm
{
    public Vm()
    {
        Values = new ObservableCollection<double>
        {
            1,
            6,
            3,
            8
        };
    }

    public ObservableCollection<double> Values { get; set; }
}
```

.. and get the same behavior (though potentially slightly lower performance as things change).

Does that work as well?

**Comment by beto-rodriguez** [comment]
@ReedCopsey Even the library could work like that, I prefer the current design, considering a couple of points:

1. I quote myself 

> ChartValues is much more than an array, or an observable collection, it is a specialized collection for the library.

This way I have more control over what is going on in a smarter way, (again it could work with ObservableCollections)

2. This class now is as portable as ObservableCollections, this project is also compiled in as a PCL so in a future we can support Xamarin and UWP (already advanced)

Any advice is appreciated

**Comment by ReedCopsey** [comment]
By requiring your library to be used, you make it impossible to use in MVVM
with existing frameworks that implement INotifyCollectionChanged
themselves, such as DynamicData, my F# lib, etc.

Ideally, you'd _require_ INCC (only), but optimize for Chart Values when
that's used.

**Comment by beto-rodriguez** [comment]
I get your point now, thanks, I reopen this and will find a solution for it

**Comment by FoggyFinder** [comment]
any updates?

**Comment by JohniMichels** [comment]
I'm actually sometimes having trouble exactly because of the code behind in order to make something that should be on View work. Cause I would really like to use in a syntax like:
```xml
<PieChart Series="{Binding DataVM}">
  <PieChart.SeriesTemplate
      Title="{Binding SerieTitle}"
      ChartValues="{Binding SerieValues}"/>
</PieChart >
```
This is assuming DataVM is a observable collection of a Type that has SerieTitle and SerieValues properties.
So this would be bound without any codebehind, and we could create the same principle for the ChartValues and make a ChartValueTemplate in order to bind the modelview custom data to the correct ChartValue Property. This way ViewModel won't need to know anything about charting, and the view will only Bind the correct data for the chart.

**Comment by Federerer** [comment]
Some time ago I've started a small project to overcome this issue:
[](https://github.com/Federerer/LiveCharts.Wpf.Extensions/tree/develop)
It needs more work and can be buggy, but works ;) Example:
```xml
<wpf:CartesianChart DisableAnimations="True" Zoom="X" DataTooltip="{x:Null}">
      <i:Interaction.Behaviors>
          <behaviors:ChartExtensionsBehavior SeriesSource="{Binding Data}">
              <behaviors:ChartExtensionsBehavior.SeriesTemplate>
                  <DataTemplate DataType="wpf:Series">
                      <wpf:LineSeries Fill="Transparent" PointGeometry="{x:Null}">
                          <i:Interaction.Behaviors>
                              <behaviors:SeriesExtensionsBehavior ValuesSource="{Binding}" XMember="TimeStamp" YMember="Value"/>
                          </i:Interaction.Behaviors>
                      </wpf:LineSeries>
                  </DataTemplate>
              </behaviors:ChartExtensionsBehavior.SeriesTemplate>
          </behaviors:ChartExtensionsBehavior>
      </i:Interaction.Behaviors>
  </wpf:CartesianChart>
```

**Comment by JohniMichels** [comment]
@Federerer Really nice, man. Are you still working on that? I just looked your git and it it looks great, going to give it a try.

**Comment by beto-rodriguez** [comment]
for version 1.0 (in progress, yes a full rewrite), `Series` and `Series.ChartValues` properties' type will be `ObservableCollection<T>` to keep the library consistent with the `.Net framework`.

**Comment by ReedCopsey** [comment]
Ideally, it'd be nice if you made them INotifyCollectionChanged, not
ObservableCollection - while ObservableCollection is the most common
implementation of this interface, the actual interface is used by WPF, etc
- and there are quite a few alternative collection implementations which
implement INotifyCollectionChanged for binding support, such as in my
libraries, DynamicData, etc.



On Thu, Dec 21, 2017 at 4:00 PM, Alberto Rodríguez <notifications@github.com

**Comment by beto-rodriguez** [comment]
@ReedCopsey the hardest part I see there, is that `INotifyCollectionChange` is not enumerable, how would I be able to plot all the points if the type is not `IEnumerable`?

Luckily `ObservableCollection<T>` is both.

**Comment by ReedCopsey** [comment]
Typically, it's handled by binding to IEnumerable, and checking at runtime for INotifyCollectionChanged.

**Comment by beto-rodriguez** [comment]
Makes sense, thanks

```c#
public abstract class Series : IViewErasableElement
    {
        private IEnumerable _values;

        public IEnumerable Values
        {
            get { return _values; }
            set
            {
                _values = value;
                var collectionChanged = _values as INotifyCollectionChanged;
                if (collectionChanged != null)
                {
                    // attach notifiers
                }
            }
        }
```

**Comment by ReedCopsey** [comment]
Yes, that's exactly the idea. Only change would be to also check the old
_values and unsubscribe before assigning, too...

**Comment by Federerer** [comment]
> Series and Series.ChartValues properties' type will be ObservableCollection<T>

Consider adding SeriesSource and SeriesTemplate properties instead of just Series. That's necessary if you want to bind a collection of viewmodels that are not aware of view types, like Series, Points etc.
